More Benchmark tests and optimizations of Shelley.Spec.Ledger.Core Relation Class

JaredCorduan opened the pull request:
This is mostly @TimSheard 's work, but I have been helping.

We now have several benchmark tests which process transactions using the `LEDGER` transition. This includes spending UTxOs, stake key registrations, deregistrations, withdrawals, stake pool registrations, re-registrations, retirements, and delegations. These tests use a variety of environment sizes (how big the UTxO set is, how many registered stake keys and stake pools), with a variety of transaction sizes (such as how many registrations, etc).

In the process, many optimizations were found and applied.

~I tried to clean up the commit history, but had a lot of trouble doing so, and abandoned it. We will soon remove the `BenchmarkCrypto` module in favor of something better, but would like to merge this long running PR fairly soon.~

closes #1504

nc6 commented on shelley/chain-and-ledger/executable-spec/bench/Main.hs:
```diff
@@ -1,16 +1,160 @@
+import Control.DeepSeq (NFData)
+import Criterion.Main -- (bench, bgroup, defaultMain, whnf)
+import Data.Word (Word64)
+-- How to precompute env for the UTxO transactions
```
Seems an odd place for comments!

JaredCorduan replied:
ah, those got rearranged by ormolu, and I didn't notice. They were comments in the import list.

nc6 commented on shelley/chain-and-ledger/executable-spec/bench/Main.hs:
```diff
@@ -1,16 +1,160 @@
+
+profileCreateRegKeys :: IO ()
+profileCreateRegKeys = do
+  putStrLn "Enter profiling stake key creation"
```
We might want to consider using `traceEventIO` here in place of `putStrLn`. Why? If we do a profiling build, `traceEventIO` adds events into the eventlog, which will allow us to more easily interpret the eventlog results.

nc6 commented on shelley/chain-and-ledger/executable-spec/bench/Main.hs:
```diff
@@ -1,16 +1,160 @@
+
+-- =========================================================
+
+varyState ::
```
Can you add some documentation here? The types give very little!

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs:
```diff
@@ -220,43 +220,48 @@ delegationTransition = do
     DCertDeleg (Delegate (Delegation hk dpool)) -> do
       -- note that pattern match is used instead of cwitness and dpool, as in the spec
-      hk ∈ dom (_stkCreds ds) ?! StakeDelegationImpossibleDELEG hk
+      haskey hk (_stkCreds ds) ?! StakeDelegationImpossibleDELEG hk
```
I would definitely be interested in investigating the use of rewrite rules here... the original syntax is much closer to the spec

TimSheard replied:
I have added some comments to indicate this.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs:
```diff
@@ -220,43 +220,48 @@ delegationTransition = do
 
-      gkh ∈ dom genDelegs ?! GenesisKeyNotInpMappingDELEG gkh
+      -- gkh ∈ dom genDelegs ?! GenesisKeyNotInpMappingDELEG gkh
+      (case Map.lookup gkh genDelegs of Just _ -> True; Nothing -> False) ?! GenesisKeyNotInpMappingDELEG gkh
```
Why not `haskey` as above?

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:
```diff
@@ -215,7 +215,7 @@ utxoInductive = do
 
   let validInputs = dom utxo
-  txins txb ⊆ validInputs ?! BadInputsUTxO (txins txb `Set.difference` validInputs)
+  all (`Map.member` v) (txins txb) ?! BadInputsUTxO (txins txb `Set.difference` validInputs)
```
This is definitely less comprehensible

nc6 commented on shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/BenchmarkFunctions.hs:
```diff
@@ -1,18 +1,56 @@
+{-# LANGUAGE DataKinds #-}
```
Do we want to move this into `bench` rather than `test`?

uroboros reviewed (approved):
this looks great to me, thanks @TimSheard

uroboros commented:
(some commits still need to be signed)

Commit:
More Benchmark tests and optimizations of Core

1) Added a few more efficient implementations of the Relation class methods
2) Changed a few places where calls to these methods were made, mostly to address inefficiences
   discovered while profiling and benchmarking a transaction that spends 1 coin, but where the
   UTxO coinset grows very large. Benchmarks seem to shrink from N-log N, to log N.
3) Adjusted benchmarks to run successfully in this branch.
4) Reworked Benchmarkfunctions.hs to use Witness sets, as Witness has been replaced by AWitness
   and RWitness.
5) Added the reward withdrawal test, which ran fine upto state of 500,000 items. Had to make
   SerializationProperties.hs work with Blake2b_256 instead of ShortHash, other wise system
   woudln't build.
6) Replaced a few calls like  (Set.singleton key ⋪ map) with (removekey key map) in
   the file Deleg.hs.
7) Left some notes about how to make range restriction (line 224) (_ptrs ds ⋫ Set.singleton hk)
   more efficient by using different data structures (emulate or use a Bimap).
8) Made some small changes to the Data.Map instance of Relation,
   mostly using more efficient Data.Map operations.
9) Removed the (Foldable f) constraint and replaced (f x) with (Set x) in the types of ◁, ▷, and ⨃.
   Then I fixed all the calls to ⨃ that were not welltypes.
   Strangely enough all the calls to ◁ and  ▷ were already Set objects
   (which is an instance of Foldable), so no changes were needed there.
10) Removed the operations (<=◁), (▷<=), (▷>=) from Core.hs, added several more efficient
    methods definitions for Map. Replaced all (x ⨃ (singleton k v)) patterns with (addpair k v x).
11) Added file BenchmarkCrypto, So that we can use Blake2b_256 hashing for benchmarking,
    but still use ConcreteCrypto (ShortHash) for testing.
12) Benchmarks also run with the Blake2b_256 hashing in the same branch.
13) Refined makebench into varyState and varyInput so we can run two kinds of tests.
    Organized the tests into two 'bgroups' first where we expect things to grow,
    and second where we expect things to remain constant.
14) A quick glance at the benchmark output HTML indicates if things are OK.

JaredCorduan commented:
@nc6 Thank you for the great review! We are going to do a follow up PR to address your comments. I wanted to get these changes merged sooner rather than later.
